File: src/csharp/pulse/cdm/io/PBActions.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;
using Google.Protobuf;

namespace Pulse.CDM
{
  public class PBAction
  {
    public static void Load(pulse.cdm.bind.ActionListData src, List<SEAction> dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.ActionListData src, List<SEAction> dst)
    {
      foreach (var action in src.AnyAction)
        dst.Add(PBAction.Load(action));
    }
    public static bool SerializeFromString(string src, List<SEAction> dst)
    {
      try
      {
        pulse.cdm.bind.ActionListData data = JsonParser.Default.Parse<pulse.cdm.bind.ActionListData>(src);
        PBAction.Load(data, dst);
      }
      catch (Google.Protobuf.InvalidJsonException)
      {
        return false;
      }
      return true;
    }

    public static pulse.cdm.bind.ActionListData Unload(List<SEAction> src)
    {
      pulse.cdm.bind.ActionListData dst = new pulse.cdm.bind.ActionListData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(List<SEAction> src, pulse.cdm.bind.ActionListData dst)
    {
      foreach (var action in src)
        dst.AnyAction.Add(PBAction.Unload(action));
    }
    public static string SerializeToString(List<SEAction> src)
    {
      var pb = PBAction.Unload(src);
      return pb.ToString();
    }

    #region AnyAction
    /** Create a new action based on the binding object,
     *  load that data into the new action, and return said action */
    public static SEAction Load(pulse.cdm.bind.AnyActionData action)
    {
      switch (action.ActionCase)
      {
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.PatientAction:
          return PBPatientAction.Load(action.PatientAction);
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.EquipmentAction:
          return PBEquipmentAction.Load(action.EquipmentAction);
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.EnvironmentAction:
            return PBEnvironmentAction.Load(action.EnvironmentAction);
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.AdvanceTime:
          {
            SEAdvanceTime adv = new SEAdvanceTime();
            PBAction.Load(action.AdvanceTime, adv);
            return adv;
          }
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.AdvanceUntilStable:
          {
            SEAdvanceUntilStable adv = new SEAdvanceUntilStable();
            PBAction.Load(action.AdvanceUntilStable, adv);
            return adv;
          }
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.SerializeRequested:
          {
            SESerializeRequested ss = new SESerializeRequested();
            PBAction.Load(action.SerializeRequested, ss);
            return ss;
          }
        case pulse.cdm.bind.AnyActionData.ActionOneofCase.SerializeState:
          {
            SESerializeState ss = new SESerializeState();
            PBAction.Load(action.SerializeState, ss);
            return ss;
          }
      }
      throw new System.InvalidOperationException("Cannot load action " + action.ToString());
    }
    /** Create a new bind object, unload the action,
     *  put that in the bind object, and return said bind object */
    public static pulse.cdm.bind.AnyActionData Unload(SEAction action)
    {
      pulse.cdm.bind.AnyActionData any = new pulse.cdm.bind.AnyActionData();
      if (action.GetType().IsSubclassOf(typeof(SEPatientAction)))
        any.PatientAction = PBPatientAction.Unload((SEPatientAction)action);
      else if (action.GetType().IsSubclassOf(typeof(SEEquipmentAction)))
        any.EquipmentAction = PBEquipmentAction.Unload((SEEquipmentAction)action);
      else if (action.GetType().IsSubclassOf(typeof(SEEnvironmentAction)))
        any.EnvironmentAction = PBEnvironmentAction.Unload((SEEnvironmentAction)action);
      else if (action is SEAdvanceTime)
        any.AdvanceTime = PBAction.Unload((SEAdvanceTime)action);
      else if (action is SEAdvanceUntilStable)
        any.AdvanceUntilStable = PBAction.Unload((SEAdvanceUntilStable)action);
      else if (action is SESerializeRequested)
        any.SerializeRequested = PBAction.Unload((SESerializeRequested)action);
      else if (action is SESerializeState)
        any.SerializeState = PBAction.Unload((SESerializeState)action);
      else
        throw new System.InvalidOperationException("Cannot unload action " + action.ToString());
      return any;
    }
    #endregion

    #region SEAction
    public static void Serialize(pulse.cdm.bind.ActionData src, SEAction dst)
    {
      dst.Clear();
      if (src != null)
      {
        if (src.Comment != null)
          dst.SetComment(src.Comment);
      }
    }

    public static void Serialize(SEAction src, pulse.cdm.bind.ActionData dst)
    {
      if (src.HasComment())
        dst.Comment = src.GetComment();
    }
    public static string SerializeToString(SEAction src)
    {
      var pb = PBAction.Unload(src);
      return pb.ToString();
    }
    #endregion

    #region SEAdvanceTime
    public static void Load(pulse.cdm.bind.AdvanceTimeData src, SEAdvanceTime dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.AdvanceTimeData src, SEAdvanceTime dst)
    {
      Serialize(src.Action, dst);
      if (src.Time != null)
        PBProperty.Load(src.Time, dst.GetTime());
    }
    public static pulse.cdm.bind.AdvanceTimeData Unload(SEAdvanceTime src)
    {
      pulse.cdm.bind.AdvanceTimeData dst = new pulse.cdm.bind.AdvanceTimeData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(SEAdvanceTime src, pulse.cdm.bind.AdvanceTimeData dst)
    {
      dst.Action = new pulse.cdm.bind.ActionData();
      Serialize(src, dst.Action);
      if (src.HasTime())
        dst.Time = PBProperty.Unload(src.GetTime());
    }
    #endregion

    #region SEAdvanceUntilStable
    public static void Load(pulse.cdm.bind.AdvanceUntilStableData src, SEAdvanceUntilStable dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.AdvanceUntilStableData src, SEAdvanceUntilStable dst)
    {
      Serialize(src.Action, dst);
      dst.SetCriteria(src.Criteria);
    }
    public static pulse.cdm.bind.AdvanceUntilStableData Unload(SEAdvanceUntilStable src)
    {
      pulse.cdm.bind.AdvanceUntilStableData dst = new pulse.cdm.bind.AdvanceUntilStableData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(SEAdvanceUntilStable src, pulse.cdm.bind.AdvanceUntilStableData dst)
    {
      dst.Action = new pulse.cdm.bind.ActionData();
      Serialize(src, dst.Action);
      dst.Criteria = src.GetCriteria();
    }
    #endregion

    #region SESerializeRequested
    public static void Load(pulse.cdm.bind.SerializeRequestedData src, SESerializeRequested dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.SerializeRequestedData src, SESerializeRequested dst)
    {
      Serialize(src.Action, dst);
      if (!string.IsNullOrEmpty(src.Filename))
        dst.SetFilename(src.Filename);
    }
    public static pulse.cdm.bind.SerializeRequestedData Unload(SESerializeRequested src)
    {
      pulse.cdm.bind.SerializeRequestedData dst = new pulse.cdm.bind.SerializeRequestedData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(SESerializeRequested src, pulse.cdm.bind.SerializeRequestedData dst)
    {
      dst.Action = new pulse.cdm.bind.ActionData();
      Serialize(src, dst.Action);
      if (src.HasFilename())
        dst.Filename = src.GetFilename();
    }
    #endregion

    #region SESerializeState
    public static void Load(pulse.cdm.bind.SerializeStateData src, SESerializeState dst)
    {
      Serialize(src, dst);
    }
    public static void Serialize(pulse.cdm.bind.SerializeStateData src, SESerializeState dst)
    {
      Serialize(src.Action, dst);
      if (!string.IsNullOrEmpty(src.Filename))
        dst.SetFilename(src.Filename);
      dst.SetType((eSerialization_Type)src.Type);
    }
    public static pulse.cdm.bind.SerializeStateData Unload(SESerializeState src)
    {
      pulse.cdm.bind.SerializeStateData dst = new pulse.cdm.bind.SerializeStateData();
      Serialize(src, dst);
      return dst;
    }
    public static void Serialize(SESerializeState src, pulse.cdm.bind.SerializeStateData dst)
    {
      dst.Action = new pulse.cdm.bind.ActionData();
      Serialize(src, dst.Action);
      if (src.HasFilename())
        dst.Filename = src.GetFilename();
      dst.Type = (pulse.cdm.bind.SerializeStateData.Types.eType)src.GetType();
    }
    #endregion
  }
}

File: src/cpp/cdm/io/protobuf/PBActions.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
PUSH_PROTO_WARNINGS
#include "pulse/cdm/bind/Actions.pb.h"
#include "pulse/cdm/bind/Scenario.pb.h"
POP_PROTO_WARNINGS
#include "cdm/io/protobuf/PBActions.h"
#include "cdm/io/protobuf/PBPatientActions.h"
#include "cdm/io/protobuf/PBEnvironmentActions.h"
#include "cdm/io/protobuf/PBEquipmentActions.h"
#include "cdm/io/protobuf/PBProperties.h"
#include "cdm/patient/actions/SEPatientAction.h"
#include "cdm/system/environment/actions/SEEnvironmentAction.h"
#include "cdm/system/equipment/SEEquipmentAction.h"
#include "cdm/engine/SEAdvanceTime.h"
#include "cdm/engine/SESerializeState.h"
#include "cdm/engine/SEOverrides.h"
#include "cdm/substance/SESubstanceManager.h"
#include "cdm/properties/SEScalarTime.h"

const std::string& eAppliedRespiratoryCycle_Name(eAppliedRespiratoryCycle c)
{
  return CDM_BIND::eAppliedRespiratoryCycle_Name((CDM_BIND::eAppliedRespiratoryCycle)c);
}

const std::string& eMergeType_Name(eMergeType m)
{
  return CDM_BIND::eMergeType_Name((CDM_BIND::eMergeType)m);
}

SEAction* PBAction::Load(const CDM_BIND::AnyActionData& action, const SESubstanceManager& subMgr)
{
  switch (action.Action_case())
  {
  case CDM_BIND::AnyActionData::kPatientAction:
    return PBPatientAction::Load(action.patientaction(), subMgr);
  case CDM_BIND::AnyActionData::kEnvironmentAction:
    return PBEnvironmentAction::Load(action.environmentaction(), subMgr);
  case CDM_BIND::AnyActionData::kEquipmentAction:
    return PBEquipmentAction::Load(action.equipmentaction(), subMgr);
  case CDM_BIND::AnyActionData::kAdvanceTime:
  {
    SEAdvanceTime* a = new SEAdvanceTime();
    PBAction::Load(action.advancetime(), *a);
    return a;
  }
  case CDM_BIND::AnyActionData::kSerialize:
  {
    SESerializeState* a = new SESerializeState();
    PBAction::Load(action.serialize(), *a);
    return a;
  }
  case CDM_BIND::AnyActionData::kOverrides:
  {
    SEOverrides* a = new SEOverrides();
    PBAction::Load(action.overrides(), *a);
    return a;
  }
  }
  subMgr.Error("Unknown Action");
  return nullptr;
}
CDM_BIND::AnyActionData* PBAction::Unload(const SEAction& action)
{
  CDM_BIND::AnyActionData* any = new CDM_BIND::AnyActionData();

  const SEAdvanceTime* adv = dynamic_cast<const SEAdvanceTime*>(&action);
  if (adv != nullptr)
  {
    any->set_allocated_advancetime(PBAction::Unload(*adv));
    return any;
  }
  const SESerializeState* ss = dynamic_cast<const SESerializeState*>(&action);
  if (ss != nullptr)
  {
    any->set_allocated_serialize(PBAction::Unload(*ss));
    return any;
  }
  const SEOverrides* o = dynamic_cast<const SEOverrides*>(&action);
  if (o != nullptr)
  {
    any->set_allocated_overrides(PBAction::Unload(*o));
    return any;
  }

  const SEPatientAction* pa = dynamic_cast<const SEPatientAction*>(&action);
  if (pa != nullptr)
  {
    any->set_allocated_patientaction(PBPatientAction::Unload(*pa));
    return any;
  }
  const SEEnvironmentAction* ea = dynamic_cast<const SEEnvironmentAction*>(&action);
  if (ea != nullptr)
  {
    any->set_allocated_environmentaction(PBEnvironmentAction::Unload(*ea));
    return any;
  }
  const SEEquipmentAction* eqa = dynamic_cast<const SEEquipmentAction*>(&action);
  if (eqa != nullptr)
  {
    any->set_allocated_equipmentaction(PBEquipmentAction::Unload(*eqa));
    return any;
  }
  action.Error("Unsupported Action");
  delete any;
  return nullptr;
}
SEAction* PBAction::Copy(const SEAction& a, const SESubstanceManager& subMgr)
{
  CDM_BIND::AnyActionData* data = PBAction::Unload(a);
  SEAction* copy = PBAction::Load(*data, subMgr);
  delete data;
  return copy;
}

void PBAction::Serialize(const CDM_BIND::ActionData& src, SEAction& dst)
{
  dst.SetComment(src.comment());
}
void PBAction::Serialize(const SEAction& src, CDM_BIND::ActionData& dst)
{
  dst.set_comment(src.m_Comment);
}

void PBAction::Load(const CDM_BIND::AdvanceTimeData& src, SEAdvanceTime& dst)
{
  dst.Clear();
  PBAction::Serialize(src, dst);
}
void PBAction::Serialize(const CDM_BIND::AdvanceTimeData& src, SEAdvanceTime& dst)
{
  PBAction::Serialize(src.action(), dst);
  if (src.has_time())
    PBProperty::Load(src.time(), dst.GetTime());
}
CDM_BIND::AdvanceTimeData* PBAction::Unload(const SEAdvanceTime& src)
{
  CDM_BIND::AdvanceTimeData* dst = new CDM_BIND::AdvanceTimeData();
  PBAction::Serialize(src, *dst);
  return dst;
}
void PBAction::Serialize(const SEAdvanceTime& src, CDM_BIND::AdvanceTimeData& dst)
{
  PBAction::Serialize(src, *dst.mutable_action());
  if (src.HasTime())
    dst.set_allocated_time(PBProperty::Unload(*src.m_Time));
}

void PBAction::Load(const CDM_BIND::SerializeStateData& src, SESerializeState& dst)
{
  dst.Clear();
  PBAction::Serialize(src, dst);
}
void PBAction::Serialize(const CDM_BIND::SerializeStateData& src, SESerializeState& dst)
{
  PBAction::Serialize(src.action(), dst);
  dst.SetType((eSerialization_Type)src.type());
  dst.SetFilename(src.filename());
}
CDM_BIND::SerializeStateData* PBAction::Unload(const SESerializeState& src)
{
  CDM_BIND::SerializeStateData* dst = new CDM_BIND::SerializeStateData();
  PBAction::Serialize(src, *dst);
  return dst;
}
void PBAction::Serialize(const SESerializeState& src, CDM_BIND::SerializeStateData& dst)
{
  PBAction::Serialize(src, *dst.mutable_action());

  dst.set_type((CDM_BIND::SerializeStateData::eType)src.m_Type);
  if (src.HasFilename())
    dst.set_filename(src.m_Filename);
}

void PBAction::Load(const CDM_BIND::OverridesData& src, SEOverrides& dst)
{
  dst.Clear();
  PBAction::Serialize(src, dst);
}
void PBAction::Serialize(const CDM_BIND::OverridesData& src, SEOverrides& dst)
{
  PBAction::Serialize(src.action(), dst);
  for (auto const& [name, o] : src.scalarproperties())
    dst.GetScalarProperties()[name] = SEScalarPair(o.value(), o.unit());
}
CDM_BIND::OverridesData* PBAction::Unload(const SEOverrides& src)
{
  CDM_BIND::OverridesData* dst = new CDM_BIND::OverridesData();
  PBAction::Serialize(src, *dst);
  return dst;
}
void PBAction::Serialize(const SEOverrides& src, CDM_BIND::OverridesData& dst)
{
  PBAction::Serialize(src, *dst.mutable_action());
  for (auto& [name, o] : src.GetScalarProperties())
  {
    CDM_BIND::ScalarData& scalar = 
      (*dst.mutable_scalarproperties())[name];
    scalar.set_value(o.value);
    scalar.set_unit(o.unit);
  }
}
void PBAction::Copy(const SEOverrides& src, SEOverrides& dst)
{
  dst.Clear();
  CDM_BIND::OverridesData data;
  PBAction::Serialize(src, data);
  PBAction::Serialize(data, dst);
}
